### ISSUE TITLE
Issue with --fmiFilter

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#12858]
@rfranke reports:

`--fmiFilter` causes issues with export of discrete states cross references to e.g. previous. Possibly also with ModelStructure. The root cause is that FMI refers to indices in the list of variables, fmiFilter reduces that list, causing variables to change their index. I need to further investigate this – but there is a workaround: `--fmiFilter=none`

Unfortunately, protected previous() variables are lost.

**Comment by casella** [comment]
@rfranke it would be nice to have some MWE that demonstrates the problem

**Comment by casella** [comment]
If this can't be fixed shortly, we can move it to 1.25.0, with the workaround `--fmiFilter=none`.

**Comment by rfranke** [comment]
Here is an example. 

```modelica
model FMIFilterTest
  input Real u;
  Integer id(start = 0);
protected
  Real p = 1;
  Real xc(start = 0);
  Real xd(start = 0);
equation
  der(xc) = u;
  when Clock() then
    xd = previous(xd) + p*u;
    id = previous(id) + 1;
  end when;
end FMIFilterTest;
```
Processed with omc calling:
```
setCommandLineOptions("--exportClocksInModelDescription"); getErrorString();
loadFile("/home/rfranke/work/FMIFilterTest.mo"); getErrorString();
buildModelFMU(FMIFilterTest, version="2.0"); getErrorString();
```
gives the following ModelVariables in modelDescription.xml:
```xml
  <ModelVariables>
  <!-- Index of variable = "1" -->
  <ScalarVariable
    name="xc"
    valueReference="0"
    initial="exact">
    <Real start="0.0"/>
  </ScalarVariable>
  <!-- Index of variable = "2" -->
  <ScalarVariable
    name="der(xc)"
    valueReference="1"
    >
    <Real derivative="1"/>
  </ScalarVariable>
  <!-- Index of variable = "3" -->
  <ScalarVariable
    name="u"
    valueReference="4"
    causality="input"
    >
    <Real start="0.0"/>
  </ScalarVariable>
  <!-- Index of variable = "4" -->
  <ScalarVariable
    name="xd"
    valueReference="5"
    variability="discrete"
    clockIndex="1"
    previous="3"
    initial="exact">
    <Real start="0.0"/>
  </ScalarVariable>
  <!-- Index of variable = "5" -->
  <ScalarVariable
    name="previous(id)"
    valueReference="0"
    variability="discrete"
    clockIndex="1"
    >
    <Integer/>
  </ScalarVariable>
  <!-- Index of variable = "6" -->
  <ScalarVariable
    name="id"
    valueReference="1"
    variability="discrete"
    clockIndex="1"
    previous="7"
    initial="exact">
    <Integer start="0"/>
  </ScalarVariable>
  </ModelVariables>
```
Following issues:
- previous(xd) is missing -- caused by default fmiFilter protected -- xd is there though
- xd has attribute previous="3", which is u -- should point to previous(xd)
- id has attribute previous="7" -- should be "5"

It works nicely for continuous-time der(xc) with attribute derivative="1" -- maybe because states appear at the top?
Note that der(xc) is not filtered out; previous(xd) is.

It looks that all trouble comes with --exportClocksInModelDescription. Maybe it can be fixed nonetheless -- or as quiick fix disable --fmiFilter with --exportClocksInModelDescription?

**Comment by rfranke** [comment]
The ModelStructure appears to be using the indices of filtered variables, e.g. 3 for u:
```
  <ModelStructure>
    <Derivatives>
      <Unknown index="2" dependencies="3" dependenciesKind="dependent" />
    </Derivatives>
    ...
  </ModelStructure>
```
@arun3688 How do you achieve this mapping from all variables to filtered variables, i.e. from 5 to 3 for u?
Can't the same be applied to the attribute previous?

**Comment by arun3688** [comment]
@rfranke previous variables should not be exported to modeldescription.xml according to the standard see the issue reported https://github.com/OpenModelica/OpenModelica/issues/11265, But when we implemented the `fmiFilter` option, we did not do proper testing on the cpp runtime or skipped it, But i need to check this, Currently i am working on another issue, i will get back to you once i found the actual problem.

**Comment by rfranke** [comment]
@adrpo, @casella, @arun3688 from my side everything is ready for 1.24 beta1. Fortunately #12854 was easy to fix.
Regarding this ticket: as the mapping of variable indices works for ModelStructure, its application to the previous attribute might be straightforward -- even though not part of the 2.0 standard, just with our 3.0 preview --exportClocksInModelDescription. Maybe already for the final 1.24?

The treatment of previous(xd) like a state when filtering, similarly to der(xc), should be covered by FMI 2.0 though. See the mentioning of previous(xd) in https://github.com/modelica/fmi-standard/blob/v2.0.x/docs/2_2_common_schema.adoc

**Comment by arun3688** [comment]
@casella @rfranke i have found the issue when exporting clocked variables and using `fmiFilter`, The issue should be fixed in this PR https://github.com/OpenModelica/OpenModelica/pull/12895, but unfortunately the server is having some issues, Hopefully it is resolved as soon as possible

**Comment by casella** [comment]
Great! Then, we can indeed include this fix in the final 1.24.0 release.

**Comment by arun3688** [comment]
@rfranke the issue is fixed with this commit d4817cddedba5a6b9d45fcfc4a4163b639bc9d35, Please test it from nightly builds tomorrow.(20-09-2024), The main issue was `fmi_index` was not used for setting the `previous` var index and also previous vars was filtered out, now previous vars are exported when used with flags `exportClocksInModelDescription`

**Comment by casella** [comment]
Great! @arun3688 please back-port this to maintenance/v1.24 and close the ticket, so that @adrpo can make the final release.

**Comment by arun3688** [comment]
@casella sure i will do that

**Comment by arun3688** [comment]
@casella I have now ported the fix to `maintenance/v1.24` branch
